pulser/simulation/simulation.py

```diff
@@ -325,55 +516,98 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
-                if initial_state.shape != (self.dim**self._size,):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = qutip.Qobj(initial_state)
-        else:
-            # by default, initial state is "ground" state of g-r basis.
-            all_ground = [self.basis['g'] for _ in range(self._size)]
-            self._initial_state = qutip.tensor(all_ground)
-
-        result = qutip.sesolve(self._hamiltonian,
-                               self._initial_state,
-                               self.eval_times,
-                               progress_bar=progress_bar,
-                               options=qutip.Options(max_step=5,
-                                                     **options)
-                               )
-        meas_basis: Optional[str]
-        if hasattr(self._seq, '_measurement'):
-            meas_basis = cast(str, self._seq._measurement)
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _run_solver(as_subroutine: bool = False,
+                        measurement_basis: str = '') -> CleanResults:
+            """Returns CleanResults: Object containing evolution results."""
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                measurement_basis = _assign_meas_basis()
+            if 'dephasing' in self.config.noise:
+                # temporary workaround due to a qutip bug when using mesolve
+                liouvillian = qutip.liouvillian(self._hamiltonian,
+                                                self._collapse_ops)
+                result = qutip.mesolve(liouvillian,
+                                       self.initial_state,
+                                       self._eval_times_array,
+                                       progress_bar=progress_bar,
+                                       options=solv_ops)
+            else:
+                result = qutip.sesolve(self._hamiltonian,
+                                       self.initial_state,
+                                       self._eval_times_array,
+                                       progress_bar=progress_bar,
+                                       options=solv_ops)
+            return CleanResults(result.states, self._size,
+                                self.basis_name, self._eval_times_array,
+                                measurement_basis)
+
+        if self.config.noise:
+            # NOISY SIMULATION:
+            meas_basis = _assign_meas_basis()
+            time_indices = range(len(self._eval_times_array))
+            total_count = np.array([Counter() for _ in time_indices])
+            # We run the system multiple times
+            for _ in range(self.config.runs):
+                # At each run, new random noise: new Hamiltonian
+                self._construct_hamiltonian()
```

Comment:
Perhaps what you want to do is to only run this if the noise in config justifies it. I believe you were trying to do this inside `_construct_hamiltonian`, but it doesn't work so well

Reply:
Well, all noises but `dephasing`justify calling this function. I should just add a check if `dephasing`is the only noise before calling it ?

Reply:
You can fine grain this more. 'SPAM' with `eta == 0` would also not need this to be called, right?

Reply:
We agreed on having it return a `CoherentResults`object when only `dephasing`is in the noise parameters.